device/transport/mqtt/src/mqtt.ts

```diff
@@ -547,4 +608,222 @@ export class Mqtt extends EventEmitter implements Client.Transport, StableConnec
 
     this._fsm.handle('getTwinReceiver', done);
   }
+
+  onDeviceMethod(methodName: string, callback: (methodRequest: DeviceMethodRequest, methodResponse: DeviceMethodResponse) => void): void {
+    this.on('method_' + methodName, callback);
+  }
+
+  private _setupSubscription(topic: TopicDescription): void {
+    debug('subscribe: ' + JSON.stringify(topic));
+    topic.subscribeInProgress = true;
```

Comment:
>topic.subscribeInProgress = true; [](start = 4, length = 33)

I realize that this is the old code.  But I get very leary of multiple booleans describing the state of a resource.  I wonder whether the case could be be to simply say subscribed = true at this point.  If the actual subscription fails then it could be set to false.  However, it would certainly seem as though the rest of the code base should be able to deal with something that says it's subscribed but then suddenly flips to false.  It could happen that way in the normal course of action, no?

Anyway, I don't expect you to change it at this point,  just something to think about.

Reply:
I believe things could indeed be simplified. and if we wanted to match AMQP we could also extract methods feature out in their own client. for now the goal is only to shuffle code around to enable the retry logic APIs. I think this feedback is worth addressing separately.

---
In reply to: [141919360](https://github.com/Azure/azure-iot-sdk-node/pull/96#discussion_r141919360) [](ancestors = 141919360)